keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,692 @@
+to certain requests before handing it off to webhooks defined using frameworks
+like Open Policy Agent.
+
+They would have to nested webhooks within the one added to the auth chain to
+have the intended effect. This enhancement allows the administrator to configure
+this behaviour via a structured API and invoke the additional webhook only when
+relevant. This also allows administrators to define `failurePolicy` behaviours for
+separate webhooks, leading to more predictable outcomes.
+
+#### Story 3: Denying requests on certain scenarios
+
+The authorizer chain should be powerful enough to deny anyone making a request
+if certain conditions are satisfied.
+
+#### Story 4: Controlling access of a privileged RBAC role
+
+Certain users associated with a privileged role might need to have their access
+scoped to certain namespaces. Having ordered authorisation modes allows the
+administrator to add a webhook restricting certain user tokens before RBAC
+grants access to the user.
+
+#### Story 5: Varying defaults across versions of the API
+
+If the default values for configuration evolve over time, affected users might
+have to override these values in case they are affected by supplying the flags
+with updated defaults. A configuration format like the one proposed can be
+versioned over time with changing defaults, mitigating this risk by only
+affecting users who knowingly bump the version.
+
+#### Story 6: Conditionally filtering requests to webhooks
+
+Let's say if an API request is being made by an user in a specific group, the
+webhook request can be skipped.
+
+
+### Risks and Mitigations
+
+- In HA clusters, there may be a skew in how the `kube-apiserver` processes in
+each are configured. This may create inconsistencies. Mitigation is to have the
+cluster bootstrapper handle such scenarios.
+- In case an administrator enables this feature and the webhook kubeconfig file
+is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
+will not be able to start. This can be mitigated by fixing the malformed values.
+
+
+## Design Details
+
+We would like to introduce a structured file format which allows authorization
+to be configured using a flag (`--authorization-config-file`) which accepts a
+path to a file on the disk. Setting both `--authorization-config-file` and
+configuring an authorization webhook will not be allowed. If the user does that,
+there will be an error and API Server would exit right away.
+
+The configuration would be validated at startup and the API server will fail to
+start if the configuration is invalid.
+
+The API server will periodically reload the configuration. If it changes, the
+new configuration will be used for the Authorizer chain. If the new configuration
+is invalid, the last known valid configuration will be used. Logging and metrics
+would be used to signal success/failure of a config reload so that cluster admins
+can have observability over this process.
```

Comment:
Add a note that the reload must not add or remove Node or RBAC authorizers... whatever the apiserver starts with is what any reloaded file must have. They can be reordered, but not added/removed.

Reply:
If the webhook is not available, then doesn't that make the new configuration invalid, which ensures the last known valid configuration will be used. Logging and metrics
can be used to signal this type of failure of a config reload so that cluster admins
can have observability over this process.

Reply:
Per discussion with @enj,
We need a time-based loop inside the apiserver to check if the file changed to trigger an update. If healthcheck on the wehobook failed, we will use the last known good config. The time-based loop will try later and when webhook health comes back, we will use the new config.

Reply:
The time-based loop alleviates my concerns with retrying non-existing webhook URLs. This way we don't need to specify a retry count with backoff mechanisms.

Adding the same to the Design Details.